Add CSS Modules support to Hydrogen

jplhomer opened the pull request:
<!-- Thank you for contributing! -->

### Description

Fixes #72 

### Additional context

This PR updates Hydrogen to add a simple Vite plugin. This plugin collects styles for each CSS Module and exports them as a `StyleTag` export.

The developer must render this style tag component somewhere in their component:

```jsx
import * as styles from './styles.module.css';

export default MyComponent() {
  return (
    <div className={styles.wrapper}>
      <styles.StyleTag />
      <p>Hello</p>
    </div>
  );
}
```

This is a v1. Eventually, we could make this smarter to collect all the styles added during RSC and potentially inject them as a single style tag or CSS file in the head, similar to how normal CSS Modules work.

This PR also adds a new example! You can play with it by running `yarn dev` in the monorepo, then `cd examples/css-modules && yarn && yarn dev`.

---

### Before submitting the PR, please make sure you do the following:

- [x] Read the [Contributing Guidelines](https://github.com/shopify/hydrogen/blob/main/docs/contributing.md)
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`)
- [x] Update docs in this repository according to your change
- [x] Run `yarn changeset add` if this PR cause a version bump based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html)

frehner commented on packages/hydrogen/src/framework/plugins/vite-plugin-css-modules-rsc.ts:
```diff
@@ -0,0 +1,37 @@
+        if (id.includes('.module.') && cssMap.has(id)) {
+          return code.replace(
+            /export default .*$/gms,
+            `import React from 'react'; export default () => React.createElement('style', {}, \`${cssMap.get(
```
Ah interesting, so it's even making a react component for you.

This does mean that it would duplicate the style tag for every time a component that uses a CSS module gets rendered, right?

Perhaps a future optimization to reduce that duplication.

frandiox replied:
This is basically rewritting the default export. I think it's generally fine, especially if we use `import * as xyz`. Another possibility would be adding a named export `HydrogenStyles` or anything like that.

frandiox commented on packages/hydrogen/src/framework/plugins/vite-plugin-css-modules-rsc.ts:
```diff
@@ -0,0 +1,37 @@
+        cssMap = new Map();
+      },
+      transform(code, id) {
+        if (id.includes('.module.')) {
```
Since this was a quick test, I didn't add a proper regexp here. Should we check for `.module.(css|scss|...)`? Or you think `.module.` won't cause other issues?

cartogram commented on examples/css-modules/package.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "example-css-modules",
+  "description": "This the default template for Hydrogen",
```
```suggestion
  "description": "An example using CSS Modules in Hydrogen",
```

cartogram commented on examples/css-modules/package.json:
```diff
@@ -0,0 +1,44 @@
+{
+  "name": "example-css-modules",
+  "description": "This the default template for Hydrogen",
+  "version": "0.11.0",
```
remove?

cartogram commented on examples/css-modules/package.json:
```diff
@@ -0,0 +1,44 @@
+    "preview": "npx @shopify/hydrogen-cli@latest preview"
+  },
+  "prettier": "@shopify/prettier-config",
+  "devDependencies": {
```
I think we should trim this to the necessary dependencies for the example.

cartogram reviewed (approved):
🎩 checks out! Nice work!

jplhomer commented:
@mcvinci Amazing thank you! I've updated our logic a bit to export an explicit `StyleTag` component. This is better, because it explains what is actually happening.

I've updated the docs in this PR to reflect this requirement. They will show up as soon as GitHub recovers from its outage 😄

jplhomer commented on packages/hydrogen/src/framework/docs/css-support.md:
```diff
@@ -82,11 +82,30 @@ If you don't want to build with Tailwind's library and instead want to write you
+{% endcodeblock %}
+
+> Tip:
+> Explore an [example implementation of CSS Modules in GitHub](https://github.com/Shopify/hydrogen/tree/v1.x-2022-07/examples/css-modules).
```
Cool. Added a ticket to support that: https://github.com/Shopify/hydrogen/issues/941 I will change my copy to point ot main

mcvinci replied:
Thank you! ❤️

blittle commented on packages/hydrogen/src/framework/plugins/vite-plugin-css-modules-rsc.ts:
```diff
@@ -0,0 +1,37 @@
+export default function cssModulesRsc() {
+  // 1. Original CSS module: `.myStyle { color: red; }`
+  // 2. CSS module after 'pre' Vite transforms: `.myStyle_hashedXYZ { color: red; }`
+  // 3. CSS module after 'post' Vite transforms: `export const myStyle = 'myStyle_hashedXYZ';`
```
I was looking at this a few days ago, and didn't understand how vite transforms work for modules. This is perfect! Thank you @frandiox !

benjaminsehl commented:
I could cry -- this happened so quickly, so beautifully. Fantastic. Let's not lose signal on the improvements, but bravo all around.

jplhomer commented:
@benjaminsehl 👍 tracking here: https://github.com/Shopify/hydrogen/issues/953
